### ISSUE TITLE
Image Score file

### ISSUE THREAD

**Issue by EstrellaXyu** [Go2Heart/EchoSight#11]
hello，thanks for your excellent work！
I want to know what is the image score? Is it the image feature or the similarity score? If it is the similarity score, It is a similarity score with whom？And if I want to get the Image score file, how can I achieve it?

**Comment by Go2Heart** [comment]
image score is the image feature generated by a given visual encoder (e.g. CLIP or EvaCLIP). If you are talking about the `scores_path` when creating your own knowledge base([the function](https://github.com/Go2Heart/EchoSight/blob/4aae05d5f063501f31c2c2f0a0187f0eb03a7188/model/retriever.py#L293)), Image score files should be files saved in pkl of python dict, with `keys` being `image_urls` and `values` being visual encoder `features` in numpy(e.g. CLIP-ViT-B14 feature: 1x768).

**Comment by EstrellaXyu** [comment]
thanks, I get it! 
I print the self.faiss_index_ids[0:71], and compare the number of "0" with the number of the image_section_indices elements of "https://en.wikipedia.org/wiki/Macrolepiota_procera"  in wiki_100_dict_v4.json. I find that the numbers of two are not equal(11 != 12), what are the reasons for this situation? 
![kb_index_ids](https://github.com/user-attachments/assets/97600f42-0457-4775-b8ec-ae23ed335053)
<img width="397" alt="wiki" src="https://github.com/user-attachments/assets/557716cb-c155-437e-a1a3-ff2de37219b3">

**Comment by Go2Heart** [comment]
This can be considered normal since EchoSight uses the repackaged pixels of [Wiki-2M](https://huggingface.co/datasets/TREC-AToMiC/AToMiC-Images-v0.2) following _Encyclopedic VQA_. Note that not all image_urls are saved in the repackaged pixel, thus EchoSight implicitly ignores these mismatches when pixels are absent in the faiss index. You can also find it in the difference between the total number of keys inside faiss_index and total number of image_urls in the knowledge base json.

**Comment by EstrellaXyu** [comment]
> This can be considered normal since EchoSight uses the repackaged pixels of [Wiki-2M](https://huggingface.co/datasets/TREC-AToMiC/AToMiC-Images-v0.2) following _Encyclopedic VQA_. Note that not all image_urls are saved in the repackaged pixel, thus EchoSight implicitly ignores these mismatches when pixels are absent in the faiss index. You can also find it in the difference between the total number of keys inside faiss_index and total number of image_urls in the knowledge base json.

Your detailed response was helpful, thank you so much!

**Comment by EstrellaXyu** [comment]
hello, I run the training script, and find that some images has no corresponding hard negative entry. Why would this happen?

**Comment by Go2Heart** [comment]
You may consider the resampling the hard negative entries. However, according to our experiments, there exists no missing of hard negative entries when training on E-VQA or InfoSeek, provided using our KB and sampled hard negative entries.

**Comment by EstrellaXyu** [comment]
> You may consider the resampling the hard negative entries. However, according to our experiments, there exists no missing of hard negative entries when training on E-VQA or InfoSeek, provided using our KB and sampled hard negative entries.

thanks for your reply! we still has some questions.
1、Could you please explain what the number '900K' refers to in terms of quantity within the context of the model training, as mentioned in your papers? Is it an arrangement of all CSV items and the corresponding "dataset_image_ids" images?
2、How long is your actual total training time? How many epochs you used in your training process?
3、During the training process, some query images do not have corresponding hard negative samples. We ran the training code and encountered the following error log:
<img width="1399" alt="bug" src="https://github.com/user-attachments/assets/6725b557-f005-4787-8edb-1d27e3be20b6" />

**Comment by Go2Heart** [comment]
1. 900K is the total IQA triplets number.
2. Total training time depends on the environment, but the training epoch around 1 or 2 should be enough.
3. Can you test again with this training csv? [link](https://drive.google.com/file/d/13BZZAserLlqKT_RHq4sX5NAIu7Ii77eE/view?usp=drive_link)

**Comment by EstrellaXyu** [comment]
> 1. 900K is the total IQA triplets number.
> 2. Total training time depends on the environment, but the training epoch around 1 or 2 should be enough.
> 3. Can you test again with this training csv? [link](https://drive.google.com/file/d/13BZZAserLlqKT_RHq4sX5NAIu7Ii77eE/view?usp=drive_link)

"Thank you very much for your patient and timely responses. We have used the new CSV file you provided, and the training code is running smoothly now! Thanks again."

**Comment by Go2Heart** [comment]
It's good to hear that. That is due to some images in the original question csv set are not included in the repackaged pixels. I shall change the provided training csv in the README. Thanks for your debugging.

**Comment by EstrellaXyu** [comment]
> It's good to hear that. That is due to some images in the original question csv set are not included in the repackaged pixels. I shall change the provided training csv in the README. Thanks for your debugging.

And could you please explain whether the reported evaluation results were obtained using the test split of E_VQA. Additionally, could you kindly provide the corresponding E_VQA evaluation CSV file for your reference?

**Comment by Go2Heart** [comment]
> > It's good to hear that. That is due to some images in the original question csv set are not included in the repackaged pixels. I shall change the provided training csv in the README. Thanks for your debugging.
> 
> And could you please explain whether the reported evaluation results were obtained using the test split of E_VQA. Additionally, could you kindly provide the corresponding E_VQA evaluation CSV file for your reference?

The test split is the same as E-VQA's.

**Comment by EstrellaXyu** [comment]
> > > It's good to hear that. That is due to some images in the original question csv set are not included in the repackaged pixels. I shall change the provided training csv in the README. Thanks for your debugging.
> > 
> > 
> > And could you please explain whether the reported evaluation results were obtained using the test split of E_VQA. Additionally, could you kindly provide the corresponding E_VQA evaluation CSV file for your reference?
> 
> The test split is the same as E-VQA's.

Thanks for your reply. I run the test script of e_VQA, and get the result below, which is inconsistent with the table 3:
<img width="1075" alt="E_VQA test results" src="https://github.com/user-attachments/assets/e6d422ce-1381-4560-a1ed-0573aca00ace" />
Here's the test script I ran, can I trouble you to see what's wrong? :
<img width="909" alt="script" src="https://github.com/user-attachments/assets/96b04638-a074-49b6-8a4b-03e7668e967b" />

**Comment by Go2Heart** [comment]
What i can tell from your result is that 1. the reranker is working that the reranked results are better; 2. the performance of the retriever doesn't meet ours, which may result in the down graded peformance overall. Maybe check the retriever module first.

**Comment by EstrellaXyu** [comment]
> What i can tell from your result is that 1. the reranker is working that the reranked results are better; 2. the performance of the retriever doesn't meet ours, which may result in the down graded peformance overall. Maybe check the retriever module first.

Thanks. And I check the code related to the retriever, which is the same as original code. I'm wondering if the performance degradation is due to erratic performance of the retriever. If it's convenient, could you provide your email for follow-up communication?

**Comment by Go2Heart** [comment]
Yes, you can find it in my website.

> > What i can tell from your result is that 1. the reranker is working that the reranked results are better; 2. the performance of the retriever doesn't meet ours, which may result in the down graded peformance overall. Maybe check the retriever module first.
> 
> Thanks. And I check the code related to the retriever, which is the same as original code. I'm wondering if the performance degradation is due to erratic performance of the retriever. If it's convenient, could you provide your email for follow-up communication?

**Comment by EstrellaXyu** [comment]
> Yes, you can find it in my website.
> 
> > > What i can tell from your result is that 1. the reranker is working that the reranked results are better; 2. the performance of the retriever doesn't meet ours, which may result in the down graded peformance overall. Maybe check the retriever module first.
> > 
> > 
> > Thanks. And I check the code related to the retriever, which is the same as original code. I'm wondering if the performance degradation is due to erratic performance of the retriever. If it's convenient, could you provide your email for follow-up communication?

好的， 我还有下面的问题请教：
1、我运行了E-VQA和infoseek的测试脚本，发现evqa的recall和vqa score和论文中的对不上，infoseek的recall是对上的，但是vqa score对不上。说明代码检索过程没有问题，怀疑是evqa对应的测试csv文件/数据库/测试脚本可能有问题，可以重新提供一下这些文件以供参考吗；
2、请问这两个数据集使用了不同评价指标（BEM/ VQA score）在代码中是如何体现的呢，我在代码中只看到了evqa的评价指标；
3、日志中报告的recall值E-VQA和infoseek都是使用的“url matching”的方式计算的吗？代码中另外一种"answer matching"的评测方式是判断问题答案是否在检索到的section中出现，其他工作也是这么算的吗？

**Comment by Go2Heart** [comment]
1. Try again with this E-VQA faiss index: [url](https://www.dropbox.com/scl/fi/75br14dx03pi13qw4qpwf/evqa_2M_faiss_index.zip?rlkey=9lehvogka4jbb7yqdvt2vpds6&st=qyksekfw&dl=0)
2. Infoseek Evaluation: [url](https://github.com/edchengg/infoseek_eval)
3. url matching is adopted due to the very same metric is done in the official paper of E-VQA. Answer matching is generally considered normal in NLP tasks, but when facing questions in E-VQA answering like "yes" or "no", issues would rise.

**Comment by EstrellaXyu** [comment]
> 1. Try again with this E-VQA faiss index: [url](https://www.dropbox.com/scl/fi/75br14dx03pi13qw4qpwf/evqa_2M_faiss_index.zip?rlkey=9lehvogka4jbb7yqdvt2vpds6&st=qyksekfw&dl=0)
> 2. Infoseek Evaluation: [url](https://github.com/edchengg/infoseek_eval)
> 3. url matching is adopted due to the very same metric is done in the official paper of E-VQA. Answer matching is generally considered normal in NLP tasks, but when facing questions in E-VQA answering like "yes" or "no", issues would rise.

Thank you very much for taking the time to reply to my question; it has resolve my questions. I will try the new faiss index.
